src/firstvoices/backend/models/characters.py

```diff
@@ -143,21 +103,123 @@ def __str__(self):
-#     """
-#
-#     class Meta:
-#         verbose_name = _("alphabet mapper")
-#         verbose_name_plural = _("alphabet mappers")
-#
-#     # See fw-4172: Once file storage is configured, enable these FileFields / change to proper paths
-#     # alphabet_mapper is a placeholder for now.
-#     # input_to_canonical = models.FileField(upload_to="alphabet_mapper")
-#     # canonical_to_base = models.FileField(upload_to="alphabet_mapper")
-#     # g2p_config = models.FileField(upload_to="alphabet_mapper")
-#
-#     def __str__(self):
-#         return self.id
+class AlphabetMapper(BaseSiteContentModel):
+    """
+    Hosts text processors and sorters for a site's custom alphabet.
+    Must be re-created upon changing any character models in an alphabet.
+    """
+
+    class Meta:
+        verbose_name = _("alphabet mapper")
+        verbose_name_plural = _("alphabet mappers")
+
+    logger = logging.getLogger(__name__)
+
+    # from all fv-character:confusables for a site
+    # JSON representation of a g2p mapping from confusable characters to canonical characters
+    input_to_canonical_map = models.JSONField(blank=True, default=dict)
+
+    @property
+    def base_characters(self):
+        """
+        Characters for the site in sort order.
+        """
+        return Character.objects.filter(site=self.site).order_by("sort_order")
+
+    @property
+    def variant_characters(self):
+        """
+        All variant characters for the site.
+        """
+        return CharacterVariant.objects.filter(site=self.site)
+
+    @property
+    def ignorable_characters(self):
+        """
+        Ignorable characters for the site.
+        """
+        return IgnoredCharacter.objects.filter(site=self.site)
+
+    @property
+    def default_g2p_config(self):
+        """
+        Returns default G2P configurations for required mappers, customized with site name and slug.
+        """
+        site_name = f"FV {self.site.title}"
+        site_code = f"fv-{self.site.slug}"
+
+        # Convert from default g2p config json to site-specific yaml
+        default_config_str = yaml.dump(
+            AppJson.objects.get(key="default_g2p_config").json
+        )
+        site_config_str = default_config_str.format(language=site_name, code=site_code)
+        site_config = yaml.safe_load(site_config_str)
+
+        return site_config
+
+    @property
+    def preprocess_transducer(self):
+        """
+        Returns an input-to-canonical G2P transducer from stored JSON map, using default config settings.
+        Does not allow manual configuration yet.
+        """
+        if self.input_to_canonical_map != {}:
+            preprocess_settings = self.default_g2p_config["preprocess_config"]
+
+            return g2p.Transducer(
+                g2p.Mapping(**preprocess_settings, mapping=self.input_to_canonical_map)
+            )
+        else:
+            self.logger.warning("No confusable map found for site %s", self.site)
+            return None
```

Comment:
Just to confirm: the sorting can still function with no preprocess transducer?

Reply:
Yes, when there is no transducer the clean confusables function of the alphabet mapper simply does not process the text and returns it.